pkg/reconciler/managed/reconciler.go

```diff
@@ -677,6 +741,13 @@ func (r *Reconciler) Reconcile(_ context.Context, req reconcile.Request) (reconc
 		managed.SetConditions(xpv1.ReconcileError(errors.Wrap(err, errReconcileConnect)))
 		return reconcile.Result{Requeue: true}, errors.Wrap(r.client.Status().Update(ctx, managed), errUpdateManagedStatus)
 	}
+	defer func() {
+		if disconnectErr := r.external.Disconnect(ctx); disconnectErr != nil {
+			log.Debug("Cannot disconnect from provider", "error", disconnectErr)
+			record.Event(managed, event.Warning(reasonCannotDisconnect, disconnectErr))
+			managed.SetConditions(xpv1.ReconcileError(errors.Wrap(disconnectErr, errReconcileDisconnect)))
```

Comment:
I think this `SetCondition` call might cause flapping/overwritten conditions in cases where the deferred function is called because we're returning from another error case where we would also update the same condition. I'll fix this up in a future PR though.